frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -594,9 +600,71 @@ GDALDataset *SENTINEL2Dataset::Open( GDALOpenInfo * poOpenInfo )
         return OpenL1C_L2A(poOpenInfo->pszFilename, SENTINEL2_L2A);
     }
 
+    if( SENTINEL2isZipped(pszHeader) )
+    {
+        CPLString osFilename(poOpenInfo->pszFilename);
+        if( strncmp(osFilename, "/vsizip/", strlen("/vsizip/")) != 0 )
+            osFilename = "/vsizip/" + osFilename;
+
+        auto psDir = VSIOpenDir(osFilename.c_str(), 1, nullptr);
+        if( psDir == nullptr ) {
+            CPLError(CE_Failure, CPLE_AppDefined, "SENTINEL2: Cannot open ZIP file %s",
+                     osFilename.c_str());
+            return nullptr;
+        }
+        while ( const VSIDIREntry* psEntry = VSIGetNextDirEntry(psDir) )
+        {
+            const char* pszInsideFilename = CPLGetFilename(psEntry->pszName);
+            if( VSI_ISREG(psEntry->nMode) && (
+                STARTS_WITH_CI(pszInsideFilename, "MTD_MSIL2A") ||
+                STARTS_WITH_CI(pszInsideFilename, "MTD_MSIL1C") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2A_OPER_MTD_SAFL1B") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2B_OPER_MTD_SAFL1B") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2A_OPER_MTD_SAFL1C") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2B_OPER_MTD_SAFL1C") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2A_USER_MTD_SAFL2A") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2B_USER_MTD_SAFL2A")) )
+            {
+                osFilename = osFilename + "/" + psEntry->pszName;
+                CPLDebug("SENTINEL2", "Trying %s", osFilename.c_str());
+                GDALOpenInfo oOpenInfo(osFilename, GA_ReadOnly);
+                VSICloseDir(psDir);
+                return Open(&oOpenInfo);
+            }
+        }
+        VSICloseDir(psDir);
+    }
+
     return nullptr;
 }
 
+/************************************************************************/
+/*                        SENTINEL2isZipped()                           */
+/************************************************************************/
+
+static bool SENTINEL2isZipped(const char* pszHeader)
```

Comment:
the value of poOpenInfo->nHeaderBytes should be passed as an extra argument ot this fi,ctop,, so it can check that its value is >= 34 + 15  (in practice the buffer should be at least 1024 bytes large, but it is dangerous to rely on this implementation detail)